File: coral-go/internal/server/routes/tasks.go
package routes

import (
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	at "github.com/cdknorow/coral/internal/agenttypes"
	"github.com/cdknorow/coral/internal/background"
	"github.com/cdknorow/coral/internal/config"
	"github.com/cdknorow/coral/internal/store"
)

// TasksHandler handles one-shot task run endpoints.
type TasksHandler struct {
	sched     *store.ScheduleStore
	cfg       *config.Config
	scheduler *background.JobScheduler
}

func NewTasksHandler(db *store.DB, cfg *config.Config) *TasksHandler {
	return &TasksHandler{
		sched: store.NewScheduleStore(db),
		cfg:   cfg,
	}
}

// SetScheduler injects the job scheduler for task launching/killing.
func (h *TasksHandler) SetScheduler(s *background.JobScheduler) {
	h.scheduler = s
}

// SubmitTask creates a one-shot task run.
// POST /api/tasks/run
func (h *TasksHandler) SubmitTask(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Prompt           string `json:"prompt"`
		RepoPath         string `json:"repo_path"`
		AgentType        string `json:"agent_type"`
		BaseBranch       string `json:"base_branch"`
		CreateWorktree   *bool  `json:"create_worktree"`
		MaxDurationS     int    `json:"max_duration_s"`
		CleanupWorktree  *bool  `json:"cleanup_worktree"`
		Flags            string `json:"flags"`
		DisplayName      string `json:"display_name"`
		WebhookURL       string `json:"webhook_url"`
		AutoAccept       bool   `json:"auto_accept"`
		MaxAutoAccepts   int    `json:"max_auto_accepts"`
	}
	if err := decodeJSON(r, &body); err != nil {
		errBadRequest(w, "invalid JSON")
		return
	}
	if strings.TrimSpace(body.Prompt) == "" {
		errBadRequest(w, "'prompt' is required")
		return
	}
	if strings.TrimSpace(body.RepoPath) == "" {
		errBadRequest(w, "'repo_path' is required")
		return
	}
	if info, err := os.Stat(body.RepoPath); err != nil || !info.IsDir() {
		errBadRequest(w, "repo_path '" + body.RepoPath + "' does not exist")
		return
	}

	// Apply defaults
	if body.AgentType == "" {
		body.AgentType = "claude"
	}
	if body.BaseBranch == "" {
		body.BaseBranch = "main"
	}
	if body.MaxDurationS == 0 {
		body.MaxDurationS = 3600
	}
	if body.MaxAutoAccepts == 0 {
		body.MaxAutoAccepts = 10
	}

	// Auto-accept flag injection (agent-type-aware)
	flags := strings.TrimSpace(body.Flags)
	if body.AutoAccept {
		var skipFlag string
		switch body.AgentType {
		case at.Codex:
			skipFlag = "--full-auto"
		case at.Gemini:
			skipFlag = "--yolo"
		default:
			skipFlag = "--dangerously-skip-permissions"
		}
		if !strings.Contains(flags, skipFlag) {
			flags = strings.TrimSpace(skipFlag + " " + flags)
		}
	}

	createWT := true
	if body.CreateWorktree != nil {
		createWT = *body.CreateWorktree
	}
	cleanupWT := true
	if body.CleanupWorktree != nil {
		cleanupWT = *body.CleanupWorktree
	}

	launchConfig := map[string]interface{}{
		"repo_path":        body.RepoPath,
		"base_branch":      body.BaseBranch,
		"agent_type":       body.AgentType,
		"prompt":           body.Prompt,
		"display_name":     body.DisplayName,
		"flags":            flags,
		"max_duration_s":   float64(body.MaxDurationS),
		"cleanup_worktree": cleanupWT,
		"create_worktree":  createWT,
		"webhook_url":      body.WebhookURL,
		"auto_accept":      body.AutoAccept,
		"max_auto_accepts": float64(body.MaxAutoAccepts),
	}

	if h.scheduler != nil {
		runID, err := h.scheduler.FireOneshot(r.Context(), launchConfig)
		if err != nil {
			if _, ok := err.(*background.ConcurrencyLimitError); ok {
				writeJSON(w, http.StatusTooManyRequests, map[string]string{"error": err.Error()})
				return
			}
			errInternalServer(w, err.Error())
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"run_id": runID, "status": "pending"})
		return
	}

	// Fallback: just create a DB record if no scheduler is wired
	now := store.NowUTC()
	var webhookURL, displayName *string
	if body.WebhookURL != "" {
		webhookURL = &body.WebhookURL
	}
	if body.DisplayName != "" {
		displayName = &body.DisplayName
	}
	runID, err := h.sched.CreateOneshotRun(r.Context(), now, displayName, webhookURL)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"run_id": runID, "status": "pending"})
}

// GetTaskStatus returns the status of a one-shot run.
// GET /api/tasks/runs/{runID}
func (h *TasksHandler) GetTaskStatus(w http.ResponseWriter, r *http.Request) {
	runID, _ := strconv.ParseInt(chi.URLParam(r, "runID"), 10, 64)
	run, err := h.sched.GetScheduledRun(r.Context(), runID)
	if err != nil {
		errNotFound(w, "run not found")
		return
	}
	writeJSON(w, http.StatusOK, run)
}

// KillTask kills a running one-shot task.
// POST /api/tasks/runs/{runID}/kill
func (h *TasksHandler) KillTask(w http.ResponseWriter, r *http.Request) {
	runID, _ := strconv.ParseInt(chi.URLParam(r, "runID"), 10, 64)

	if h.scheduler != nil {
		if !h.scheduler.KillRun(r.Context(), runID) {
			errNotFound(w, "Run not found or not active")
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"ok": true})
		return
	}

	// Fallback: just update DB
	run, err := h.sched.GetScheduledRun(r.Context(), runID)
	if err != nil || run == nil {
		errNotFound(w, "Run not found or not active")
		return
	}
	h.sched.UpdateScheduledRun(r.Context(), runID, map[string]interface{}{
		"status":      "killed",
		"exit_reason": "user_cancelled",
	})
	writeJSON(w, http.StatusOK, map[string]any{"ok": true})
}

// ListTasks returns recent one-shot runs.
// GET /api/tasks/runs
func (h *TasksHandler) ListTasks(w http.ResponseWriter, r *http.Request) {
	limit := queryInt(r, "limit", 50)
	if limit > 200 {
		limit = 200
	}
	status := r.URL.Query().Get("status")
	var statusPtr *string
	if status != "" {
		statusPtr = &status
	}
	runs, err := h.sched.ListOneshotRuns(r.Context(), limit, statusPtr)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"runs": emptyIfNil(runs)})
}

// ListActiveRuns returns currently running tasks.
// GET /api/tasks/active
func (h *TasksHandler) ListActiveRuns(w http.ResponseWriter, r *http.Request) {
	runs, err := h.sched.ListActiveRuns(r.Context())
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"runs": emptyIfNil(runs)})
}

File: coral-python/coral/api/tasks.py
"""REST API for one-shot task runs (Jobs)."""

from __future__ import annotations

import os
from pathlib import Path
from typing import TYPE_CHECKING

from fastapi import APIRouter, Query

if TYPE_CHECKING:
    from coral.background_tasks.scheduler import JobScheduler
    from coral.store.schedule import ScheduleStore

router = APIRouter(prefix="/api/tasks", tags=["Tasks"])

# Module-level dependencies, set by web_server.py
store: ScheduleStore = None  # type: ignore[assignment]
scheduler: JobScheduler = None  # type: ignore[assignment]


@router.post("/run")
async def submit_task(body: dict):
    """Submit a one-shot agent task."""
    from coral.background_tasks.scheduler import ConcurrencyLimitError

    prompt = body.get("prompt", "").strip()
    if not prompt:
        return {"error": "'prompt' is required"}, 400

    repo_path = body.get("repo_path", "").strip()
    if not repo_path:
        return {"error": "'repo_path' is required"}, 400

    if not Path(repo_path).is_dir():
        return {"error": f"repo_path '{repo_path}' does not exist"}, 400

    flags = body.get("flags", "")
    if body.get("auto_accept", False):
        skip_flag = "--dangerously-skip-permissions"
        if skip_flag not in flags:
            flags = f"{skip_flag} {flags}".strip()

    webhook_url = body.get("webhook_url")
    if webhook_url:
        from coral.api.webhooks import _validate_url
        url_error = _validate_url(webhook_url, "generic")
        if url_error:
            return {"error": f"Invalid webhook_url: {url_error}"}, 400

    config = {
        "prompt": prompt,
        "repo_path": repo_path,
        "agent_type": body.get("agent_type", "claude"),
        "base_branch": body.get("base_branch", "main"),
        "create_worktree": body.get("create_worktree", True),
        "max_duration_s": body.get("max_duration_s", 3600),
        "cleanup_worktree": body.get("cleanup_worktree", True),
        "flags": flags,
        "display_name": body.get("display_name"),
        "webhook_url": webhook_url,
        "auto_accept": body.get("auto_accept", False),
        "max_auto_accepts": body.get("max_auto_accepts", 10),
    }

    try:
        run_id = await scheduler.fire_oneshot(config)
    except ConcurrencyLimitError as e:
        return {"error": str(e)}, 429

    return {"run_id": run_id, "status": "pending"}


@router.get("/runs/{run_id}")
async def get_run(run_id: int):
    """Poll status of a specific run."""
    run = await store.get_scheduled_run(run_id)
    if not run:
        return {"error": "Run not found"}, 404
    return {
        "id": run["id"],
        "status": run["status"],
        "trigger_type": run.get("trigger_type", "cron"),
        "session_id": run.get("session_id"),
        "display_name": run.get("display_name"),
        "worktree_path": run.get("worktree_path"),
        "scheduled_at": run.get("scheduled_at"),
        "started_at": run.get("started_at"),
        "finished_at": run.get("finished_at"),
        "exit_reason": run.get("exit_reason"),
        "error_msg": run.get("error_msg"),
    }


@router.post("/runs/{run_id}/kill")
async def kill_run(run_id: int):
    """Cancel a running or pending task."""
    ok = await scheduler.kill_run(run_id)
    if not ok:
        return {"error": "Run not found or not active"}, 404
    return {"ok": True}


@router.get("/runs")
async def list_runs(limit: int = Query(50, ge=1, le=200), status: str | None = None):
    """List recent one-shot task runs."""
    runs = await store.list_oneshot_runs(limit=limit, status=status)
    return {"runs": runs}


@router.get("/active")
async def list_active_runs():
    """List all active runs (both cron and API-triggered)."""
    runs = await store.list_active_runs()
    return {"runs": runs}
